### PR TITLE
Substrate Integrations

### PR DESCRIPTION
**Pull request by dOrgJelli**

### CODE REVIEW

**Review by dOrgJelli — COMMENTED** [review]
Okay I've gone through everything pretty thoroughly except the examples. My final point of feedback is a bit more macro level.

I think we should rename this wrapper to `substrate-rpc-wrapper` instead of `substrate-core-wrapper`. Core is used in the Substrate docs to refer to the node's core runtime:
https://substrate-developer-hub.github.io/docs/en/

Also I think we should model the layout of this project after an existing RPC SDK. I found this really great one in Go: https://pkg.go.dev/github.com/centrifuge/go-substrate-rpc-client

If we do this, the logic could be organized in the following way (roughly speaking):
```
author
- pendingExtrinsics
- submitAndWatchExtrinsic
- submitExtrinsic
chain
- getBlock
- getBlockHash
- getBlockHashLatest
- getBlockLatest
- getFinalizedHead
- ...etc
state
- getStorage
- getStorageHash
- getChildStorage
- getChildStorageHash
- ...etc
system
- chain
- health
- name
- networkState
- ...etc
```
I'm not the expert here, so would like to have a conversation about this. Want to make sure we're organizing things in the most logic way for developers who are used to other existing SDKs.

File: protocol/substrate/core-wrapper/module/schema.graphql [hunk 1]
```diff
@@ -0,0 +1,69 @@
+#import { Module, Request, Header, UrlParam, Response, } into Http from "wrap://ens/http.polywrap.eth"
+
+type Module {
```
[the remark this answers is not in the document]
Comment (dOrgJelli) [reply protocol/substrate/core-wrapper/module/schema.graphql:3]: I know that these were in the W3F proposal, and we can hold true to these, but I think that in this wrapper's final form we should adhere to all of the functionality exposed by similar SDKs (like the Go SDK referenced in my final comment of this review).

File: protocol/substrate/core-wrapper/module/src/types/account_info.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    /// The amount that `free` may not drop below when withdrawing for *anything except transaction
+    /// fee payment*.
+    pub misc_frozen: Balance,
+    /// The amount that `free` may not drop below when withdrawing specifically for transaction
+    /// fee payment.
+    pub fee_frozen: Balance,
+}
+
+/// Type used to encode the number of references an account has.
+pub type RefCount = u32;
```
[the remark this answers is not in the document]
Comment (ivanceras) [reply protocol/substrate/core-wrapper/module/src/types/account_info.rs]: There are unused types and structs that was originally from the `substrate-api-client` where I copied most of the code.  When doing some refactoring and function improvements, I realized some of the complex types are not needed anymore. I will remove those unused types as I'm doing the finishing touches.

File: protocol/substrate/core-wrapper/module/src/types/extrinsic_params.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+impl From<PlainTip> for u128 {
+    fn from(tip: PlainTip) -> Self {
+        tip.tip
+    }
+}
+
+/// A tip payment made in the form of a specific asset.
+#[derive(Copy, Clone, Debug, Default, Decode, Encode, Eq, PartialEq)]
+pub struct AssetTip {
```
[the remark this answers is not in the document]
Comment (ivanceras) [reply protocol/substrate/core-wrapper/module/src/types/extrinsic_params.rs]: This one is not used anymore, as I've found a simpler way to do this.

File: protocol/substrate/core-wrapper/module/src/api/storage_api.rs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    Encode,
+};
+use scale_info::{
+    form::PortableForm,
+    Type,
+};
+use sp_core::storage::StorageKey;
+
+impl Api {
+    // curl -H "Content-Type: application/json" -d '{"id":"1","jsonrpc":"2.0","method":"state_getStorage","params":["0x26aa394eea5630e07c48ae0c9558cef7b99d880ec681799c0cf30e8886371da9de1e86a9a8c739864cf3cc5ec2bea59fd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d",null]}' http://localhost:9933/
```
[the remark this answers is not in the document]
Comment (ivanceras) [reply protocol/substrate/core-wrapper/module/src/api/storage_api.rs]: These are some really old notes when trying out the rpc from cli, this will be removed in the final code.

File: protocol/substrate/core-wrapper/module/src/api/base_api.rs [hunk 5]
```diff
@@ -0,0 +1,303 @@
+use crate::{
```
[the remark this answers is not in the document]
Comment (ivanceras) [reply protocol/substrate/core-wrapper/module/src/api/base_api.rs]: The `BaseApi` contains some straightforward 1 http call to get the result, where us the `Api` requires fetching some initial data such as metadata and runtime version.
 I did tried to design to do fetch it lazily, but the added complexity of doing it, would still not be optimal and with a more ugly code. This should be done in optimization phase such as doing storing a pool of reusable Api for each substrate url node as the key which ultimately fetch the required parameters exactly only once for each chain at the specified url.

File: protocol/substrate/core-wrapper/tests/mutation.spec.ts [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        args: {
+          url: "http://localhost:9933",
+          number: 0
+        }
+    });
+
+    // Or use the test app's codegen types (see web3api.app.yaml)
+    const result = await Substrate_Module.blockHash(
+      {
+          url: "http://localhost:9933",
```
[the remark this answers is not in the document]
Comment (dOrgJelli) [reply protocol/substrate/core-wrapper/tests/mutation.spec.ts]: Resolved here: https://github.com/ChainSafe/integrations/pull/5

**Comment by dOrgJelli** [comment]
Hey @ivanceras I saw you added docs here, but would you mind adding them to the schema since that is what external developers will most likely see when interacting with the wrapper?
https://github.com/polywrap/integrations/pull/97/commits/e2d5e222c9cdf6993ae3f31c597fbb14bbfa6626

An example of this can be found here: https://spec.graphql.org/June2018/#sec-Descriptions

```graphql
type Module {
  """
  multi-line comment
  """
  method(
    arg: String!
  ): String!
}
```

**Comment by dOrgJelli** [comment]
Also @ivanceras, I noticed when running tests w/o any nodes running locally (on localhost), methods were returning `null` in the `result.data`, instead of returning an error in the `result.error` property.

Could we please add tests that check for proper error messaging? Currently the tests only test for happy paths and basic input / output validation.

**Comment by willemolding** [comment]
I would like to propose marking this as a non-draft PR and opening it up to a final review and run of CI before merging
